### PR TITLE
[refactor] Remove dependency on get_current_program() in ui/backends/vulkan

### PR DESCRIPTION
**Pull request by PGZXB**
Related issue = #3955

### CODE REVIEW

File: python/taichi/ui/window.py [hunk 1]
```diff
@@ -16,12 +16,13 @@ class Window(_ti_core.PyWindow):
         res (Tuple[Int]): resolution (width, height) of the window, in pixels.
         layout (vsync): whether or not vertical sync should be enabled.
     """
-    def __init__(self, name, res, vsync=False, show_window=True):
+    def __init__(self, name, res, vsync=False, show_window=True, prog=None):
```
Comment (k-ye) [inline python/taichi/ui/window.py:19]: Thanks! I think it's fine to use `get_runtime().prog` at Python level. The CPP part needs to get rid of these global states.

**Review by k-ye — APPROVED** [review]
LGTM + nit!

**Comment by AmesingFlank** [comment]
Let's make the `Program*` a member of `AppContext`, which can be accessed by all the renderer/renderable classes.

File: taichi/ui/backends/vulkan/canvas.h [hunk 2]
```diff
@@ -37,7 +37,7 @@ namespace vulkan {
 
 class Canvas final : public CanvasBase {
  public:
-  Canvas(Renderer *renderer);
+  Canvas(lang::Program *prog, Renderer *renderer);
```
[the remark this answers is not in the document]
Comment (PGZXB) [reply taichi/ui/backends/vulkan/canvas.h]: > Do we need this argument here?

Sorry, my bad. This cause a compile error. I have already remove it. I don't have computer which can build Taichi with Vulkan now, so I cannot verify.

[the remark this answers is not in the document]
Comment (AmesingFlank) [reply taichi/ui/backends/vulkan/canvas.h]: No worries. Let's see if CI passes. By the way, you don't need a vulkan-capable GPU to build taichi with Vulkan.

**Pull request merged**